commit fa09d2d683079d316286b27452edff2b8fe57fde (refs/changes/93/3993/1)
Author: Pedro Palazon Candel <pedro@joyent.com>
Date:   2018-05-23T21:08:41+02:00 (1 year, 5 months ago)
    
    TRITON-325 sdcadm still aggressively gives up after updating non-HA moray

diff --git a/lib/procedures/update-moray-v2.js b/lib/procedures/update-moray-v2.js
index 6a83962..237d1d9 100644
--- a/lib/procedures/update-moray-v2.js
+++ b/lib/procedures/update-moray-v2.js
@@ -425,7 +425,9 @@ UpdateMorayV2.prototype.execute = function morayv2Execute(opts, cb) {
                         alias: ctx.change.inst.alias,
                         server: ctx.change.inst.server,
                         domain: ctx.change.service.metadata.SERVICE_DOMAIN,
-                        sdcadm: opts.sdcadm
+                        sdcadm: opts.sdcadm,
+                        progress: progress,
+                        log: opts.log
                     }, next);
                 },
                 function waitUntilTmpVMNotInDNS(ctx, next) {
@@ -487,10 +489,9 @@ UpdateMorayV2.prototype.execute = function morayv2Execute(opts, cb) {
             if (change.insts && change.insts.length === 1 && !change.inst) {
                 change.inst = change.insts[0];
             }
-
             // Having a temporary instance doesn't mean we're on HA setup:
             if ((insts.length > 1 && !tmpInsts.length) ||
-                (insts.length >= tmpInsts.length + 1)) {
+                (tmpInsts.length && (insts.length > tmpInsts.length + 1))) {
                 change.HA = true;
             }
 
diff --git a/test/update.test.js b/test/update.test.js
index 1fd5b84..d3ebd15 100644
--- a/test/update.test.js
+++ b/test/update.test.js
@@ -189,6 +189,34 @@ test('sdcadm update --force-same-image', function (t) {
     });
 });
 
+// We've had several issues in the past regarding SAPI's moray client being
+// unable to recover from connection lost when we update sapi right after
+// non-HA moray. Let's add a test to check if we hit any more issues:
+test('update non-HA moray and SAPI consecutively', function (t) {
+    var cmd = 'sdcadm up moray sapi --force-same-image --yes';
+    exec(cmd, function (err, stdout, stderr) {
+        t.ifError(err, 'Execution error');
+        t.equal(stderr, '', 'Empty stderr');
+
+        var findStrings = [
+            'Updating moray',
+            'Provisioning Temporary moray',
+            'Reprovisioning VM',
+            'Destroying tmp VM',
+            'Updating sapi',
+            'Provisioning Temporary sapi',
+            'Reprovisioning sapi VM',
+            'Stop tmp VM',
+            'Updated successfully'
+        ];
+
+        findStrings.forEach(function (str) {
+            t.notEqual(stdout.indexOf(str), -1,
+                    util.format('check update string present %s', str));
+        });
+        t.end();
+    });
+});
 
 // As part of teardown, we'll not only rollback the updates, but also remove
 // the images we imported, since this is the only way to test the whole
